users/vieting/experiments/switchboard/ctc/feat/network_helpers/specaug_configurable.py

```diff
@@ -0,0 +1,300 @@
+                pos=indices[:, i],
+                max_amount=max_dims,
+                sorted_indices=sorted_indices,
+            )
+    else:
+        _, x = tf.while_loop(
+            cond=lambda i, _: tf.less(i, tf.reduce_max(num)),
+            body=lambda i, x: (
+                i + 1,
+                tf.where(
+                    tf.expand_dims(tf.expand_dims(tf.less(i, num), axis=-1), axis=-1),
+                    _mask(
+                        x,
+                        batch_axis=batch_axis,
+                        axis=axis,
+                        pos=indices[:, i],
+                        max_amount=max_dims,
+                        sorted_indices=sorted_indices,
+                    ),
+                    x,
+                ),
+            ),
+            loop_vars=(0, x),
+        )
+    return x
+
+
+def transform(data, network, **config):
+    import tensorflow as tf
+
+    x = data.placeholder
+    step = network.global_train_step
+    current_epoch = tf.cast(step / config["steps_per_epoch"], tf.int32)
+
+    specaug_params = config["specaug_params"]
+
+    # Determine if we should use sorting
+    use_sorting = tf.logical_and(
+        config["enable_sorting"], tf.greater_equal(current_epoch, config["sorting_start_epoch"])
+    )
+
+    # Get filter indices (sorted or unsorted)
+    filter_layer = network.layers["features"].subnetwork.layers["conv_h_filter"].output.placeholder
+    num_filters = network.layers["features"].subnetwork.layers["conv_l"].output.shape[-1]
+
+    def get_sorted_indices():
+        sorted_filter_indices = sort_filters_by_center_freq(filter_layer)
+        sorted_indices = tf.stack(
+            [sorted_filter_indices * num_filters + filter_idx for filter_idx in range(num_filters)]
+        )
+        return tf.reshape(tf.transpose(sorted_indices), (-1,))
+
+    def get_unsorted_indices():
+        return tf.range(tf.shape(x)[data.feature_dim_axis])
+
+    sorted_indices = tf.cond(use_sorting, get_sorted_indices, get_unsorted_indices)
+
+    def get_masked():
+        x_masked = x
+        time_mask_max_size = tf.gather(specaug_params["time_mask_max_size"], current_epoch)
+        time_mask_max_num = tf.gather(specaug_params["time_mask_max_num"], current_epoch)
+        freq_mask_max_num = tf.gather(specaug_params["freq_mask_max_num"], current_epoch)
+        freq_mask_max_size = tf.gather(specaug_params["freq_mask_max_size"], current_epoch)
+        time_mask_max_proportion = tf.cast(
+            tf.math.floor(config["time_mask_max_proportion"] * tf.cast(tf.shape(x)[data.time_dim_axis], tf.float32)),
+            tf.int32,
+        )
+        freq_mask_max_proportion = tf.cast(
+            tf.math.floor(config["freq_mask_max_proportion"] * tf.cast(tf.shape(x)[data.feature_dim_axis], tf.float32)),
+            tf.int32,
+        )
+
+        # check for the limits
+        actual_time_mask_max_num = tf.minimum(time_mask_max_num, time_mask_max_proportion // time_mask_max_size)
```

Comment:
I wonder if this makes sense. If you have max proportion of 0.5 and a sequence length of 100, you want to mask at most 50 frames. The default max mask size is 10. However, you cannot tell the individual mask sizes here. You would get an actual max number of 5. This is somewhat pessimistic since the actual masks could be smaller resulting in less masked frames than intended.

Reply:
time_mask_max_proportion is intended as a safety check to ensure the scheduling is not to aggressive. It should not be reached if scheduling,, base_values, steps per epoch, num_epochs, and the number of features is all known and set correctly. I tried to do it as a assertion first, but it is not possible to build assertion into the tf graph.